src/org/opensolaris/opengrok/search/context/FormattedLines.java

```diff
@@ -0,0 +1,142 @@
+ * result for presentation.
+ */
+public class FormattedLines {
+    private final SortedMap<Integer, String> lines = new TreeMap<>();
+    private String footer;
+    private boolean limited;
+
+    /**
+     * Gets a count of the number of lines in the instance.
+     */
+    public int getCount() {
+        return lines.size();
+    }
+
+    /**
+     * @return the footer
+     */
+    public String getFooter() {
+        return footer;
+    }
+
+    /**
+     * @param footer the footer to set
+     */
+    public void setFooter(String footer) {
+        this.footer = footer;
+    }
+
+    /**
+     * Gets a value indicating if lines were limited.
+     * @return the value
+     */
+    public boolean isLimited() {
+        return limited;
+    }
+
+    /**
+     * Sets a value indicating if lines were limited.
+     * @param limited the value
+     */
+    public void setLimited(boolean limited) {
+        this.limited = limited;
+    }
+
+    /**
+     * Removes the highest line from the instance.
+     * @return a defined value
+     * @throws NoSuchElementException if the instance is empty
+     */
+    public String pop() {
+        String value = lines.remove(lines.lastKey());
+        return value;
+    }
+
+    /**
+     * Sets the specified String line for the specified line number, replacing
+     * any previous entry for the same line number.
+     * @param lineno a value
+     * @param line a defined instance
+     * @return the former value or {@code null}
+     */
+    public String put(int lineno, String line) {
+        if (line == null) {
+            throw new IllegalArgumentException("`line' is null");
+        }
+        return lines.put(lineno, line);
+    }
+
+    /**
+     * Creates a new instance with lines merged from this instance and
+     * {@code other}. Any lines in common for the same line number are taken
+     * from this instance rather than {@code other}; and likewise for
+     * {@link #getFooter()}.
+     * <p>
+     * {@link #isLimited()} will be {@code true} if either is {@code true}, but
+     * the value is suspect since it cannot be truly known if the merged result
+     * is actually the unlimited result.
+     * @param other a defined instance
+     * @return a defined instance
+     */
+    public FormattedLines merge(FormattedLines other) {
+        FormattedLines res = new FormattedLines();
+        res.lines.putAll(this.lines);
+        for (Map.Entry<Integer, String> kv : other.lines.entrySet()) {
+            res.lines.putIfAbsent(kv.getKey(), kv.getValue());
+        }
+
+        res.setFooter(this.footer != null ? this.footer : other.footer);
+        res.setLimited(this.limited || other.limited);
+        return res;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder bld = new StringBuilder();
+        for (String line : lines.values()) {
+            bld.append(line);
+        }
+        String f = footer;
```

Comment:
Maybe remove the variable `f`? I think it is nicer to have it just as:
```
if (footer != null && limited) {
    bld.append(footer);
}
```

Reply:
if there are concerns about race condition could it be solved otherwise?
e.g. sync on setter?
I was under the impression this object would be thread local when used ...

Reply:
Well my point is that my making a local copy is rationalized — i.e. it's a pattern I do recommend generally when nullably-mutable fields are involved so that you can have objects that do not use heavyweight synchronization but also that are not inconsistent intra-method.

A local copy is also valuable in some synchronized settings — e.g., where you have field data that strictly evolves lazily from `null` to non-`null` (as in e.g. a singleton) and a `volatile` field is properly used. In that case too, it's better to make a local copy of the field so that for the predominant case happening after initialization (i.e., when non-`null`), only a single atomic read is required w.r.t. to field.

Reply:
ok, makes sense